python/ql/src/semmle/python/objects/TObject.qll

```diff
@@ -0,0 +1,466 @@
+import python
+private import semmle.python.types.Builtins
+private import semmle.python.objects.ObjectInternal
+private import semmle.python.pointsto.PointsTo
+private import semmle.python.pointsto.PointsToContext
+
+newtype TObject =
+    TBuiltinClassObject(Builtin bltn) {
+        bltn.isClass() and
+        not bltn = Builtin::unknownType() and
+        not bltn = Builtin::special("type")
+    }
+    or
+    TBuiltinFunctionObject(Builtin bltn) { bltn.isFunction() }
+    or
+    TBuiltinMethodObject(Builtin bltn) { bltn.isMethod() }
+    or
+    TBuiltinModuleObject(Builtin bltn) { bltn.isModule() }
+    or
+    TBuiltinOpaqueObject(Builtin bltn) {
+        not bltn.isClass() and not bltn.isFunction() and
+        not bltn.isMethod() and not bltn.isModule() and
+        not bltn.getClass() = Builtin::special("tuple") and
+        not exists(bltn.intValue()) and
+        not exists(bltn.floatValue()) and
+        not exists(bltn.strValue()) and
+        not py_special_objects(bltn, _)
+    }
+    or
+    TPythonFunctionObject(ControlFlowNode callable) {
+        callable.getNode() instanceof CallableExpr
+    }
+    or
+    TPythonClassObject(ControlFlowNode classexpr) {
+        classexpr.getNode() instanceof ClassExpr
+    }
+    or
+    TPackageObject(Folder f) {
+        exists(moduleNameFromFile(f))
+    }
+    or
+    TPythonModule(Module m) {
+        not m.isPackage() and not exists(SyntaxError se | se.getFile() = m.getFile())
+    }
+    or
+    TTrue()
+    or
+    TFalse()
+    or
+    TNone()
+    or
+    TUnknown()
+    or
+    TUnknownClass()
+    or
+    TUndefined()
+    or
+    TInt(int n) {
+        // Powers of 2 are used for flags
+        is_power_2(n) or
+        // And all combinations of flags up to 2^8
+        n in [0..511] or
```

Comment:
We could do this, but it is for specific cases not general analysis.
Mainly it is so we know the value of `x = re.MULTILINE | re.VERBOSE` and that `x & re.VERBOSE` is true to avoid FPs in queries involving regexes.